src/kcas.ml

```diff
@@ -10,69 +10,102 @@ module Id = struct
-
-(* Note: The writes to `state.casn <- ...` above could be removed to reduce time
-   at the cost of increasing space usage (by a constant factor). *)
-
-let release cass = function
-  | `After -> release_after cass
-  | `Before -> release_before cass
-
-let finish casn (`Undetermined cass as undetermined)
-    (status : [ `Before | `After ]) =
+      release_before casn lt |> ignore;
+      if not (is_cmp casn state) then (
+        state.after <- state.before;
+        state.casn <- casn_before);
+      release_before casn gt
+
+let release casn cass = function
+  | `After -> release_after casn cass
+  | `Before -> release_before casn cass
+
+let rec verify casn = function
+  | NIL -> `After
+  | CASN (atom, desired, lt, gt) -> (
+      match verify casn lt with
+      | `After ->
+          if is_cmp casn desired && Atomic.get atom.state != desired then
+            `Before
+          else verify casn gt
+      | status -> status)
+
+let finish casn (`Undetermined cass as undetermined) (status : determined) =
   if Atomic.compare_and_set casn (undetermined :> status) (status :> status)
-  then release cass status
+  then release casn cass status
   else Atomic.get casn == `After
 
-let rec determine casn undetermined skip = function
-  | NIL -> skip || finish casn undetermined `After
-  | CASN (loc, desired, lt, gt) as eq ->
-      determine casn undetermined true lt
-      &&
-      let current = Atomic.get loc.state in
-      if desired == current then determine casn undetermined skip gt
-      else
-        let expected = desired.before in
-        if
-          expected == current.after
-          && (current.casn == casn_after || is_after current.casn)
-          || expected == current.before
-             && (current.casn == casn_before || not (is_after current.casn))
-        then
-          let status = Atomic.get casn in
-          if status != (undetermined :> status) then status == `After
-          else if Atomic.compare_and_set loc.state current desired then
-            determine casn undetermined skip gt
-          else determine casn undetermined skip eq
-        else finish casn undetermined `Before
+let exit _ = raise Exit [@@inline never]
+
+let rec determine casn action = function
+  | NIL -> action
+  | CASN (loc, desired, lt, gt) as eq -> (
+      match determine casn action lt with
+      | `Before -> `Before
+      | action ->
+          let current = Atomic.get loc.state in
+          if desired == current then
+            let action = if is_cmp casn desired then `Verify else action in
+            determine casn action gt
+          else
+            let expected = desired.before in
+            if
+              (not (is_cmp casn desired))
+              && (expected == current.after
+                  && (current.casn == casn_after || is_after current.casn)
+                 || expected == current.before
+                    && (current.casn == casn_before
+                       || not (is_after current.casn)))
```

Comment:
This condition is getting complicated; perhaps it can be divided into components with meaningful names?

Reply:
I agree, the conditional here is hairy.  The annoying problem here (again) is that that the existential in the GADT prevents some refactorings (or at least makes the harder/more expensive (I mean, we could avoid all the magic using Leibniz equality, but I wouldn't go there)), but I moved the original conditional to a local helper function and perhaps the main intent is now clearer.